Make `CipherCtx::cipher_update` more flexible

wiktor-k opened the pull request:
This change relaxes constraints on the output buffer size when it can be safely determined how many bytes will be put in the output buffer.

For supported cryptographic backends (OpenSSL >= 1.1) the cipher's `num` parameter will be consulted for the number of bytes in the block cache. For unsupported backends the behavior will not change (the code will assume full block in the cache).

For callers that do the check themselves and want to use other backends (e.g. BoringSSL or LibreSSL) and unsafe `unchecked_cipher_update` function is added.

Additionally a `CipherCtx::minimal_output_size` function is added for letting the callers know how big should the output buffer be for the next `cipher_update` call.

Fixes #1729.

See: https://mta.openssl.org/pipermail/openssl-users/2022-November/015623.html

sfackler commented on openssl/src/cipher_ctx.rs:
```diff
@@ -363,6 +363,65 @@ impl CipherCtxRef {
+    /// # Panics
+    ///
+    /// Panics if the context has not been initialized with a cipher.
+    pub fn minimal_output_size(&self, inlen: usize) -> usize {
```
I usually don't like to add extra methods on top of OpenSSL APIs, but it seems like it may be reasonable to make an exception here.

wiktor-k replied:
Understood!

sfackler commented on openssl/src/cipher_ctx.rs:
```diff
@@ -501,33 +560,61 @@ impl CipherCtxRef {
+    /// buffer depends on the state of partially updated blocks (see
+    /// [`Self::minimal_output_size`]).
+    #[corresponds(EVP_CipherUpdate)]
+    pub unsafe fn unchecked_cipher_update(
```
Nit: I believe the Rust naming convention for this would be `cipher_update_unchecked` to match e.g. `[T]::get_unchecked`.

wiktor-k replied:
Yep, good catch! Thanks.

sfackler commented:
Looks good to me other than the method naming!

wiktor-k commented:
Rebased on top of `master` and fixed the naming issue.

Thanks for your reviews! I really appreciate it :pray: 

:thinking: It seems I missed a release with this PR... if you don't mind I'll file a couple of small documentation + example PRs.

sfackler commented:
PRs are always welcome!

sfackler commented:
I'm happy to cut a new release as well if you'd like.

wiktor-k commented:
> I'm happy to cut a new release as well if you'd like.

Thank you! I'll be back with one more documentation PR and then I'll kindly ask you for a new release (it will also look better on docs.rs). Thanks in advance, I'll be back :)
